### ISSUE TITLE
Better multiprocessing dataloader support for large datasets

### ISSUE THREAD

**Issue by eric-haibin-lin** [dmlc/gluon-nlp#707]
## problem statement
I have a large corpus for training, and ideally I don't want to load them all at once due to limited CPU memory. Currently, the `DatasetStream` API serves this purpose by returning one Dataset at a time. 

However, the `DatasetStream.__iter__` API limits the data pipeline performance if the data preprocessing is complicated and computational heavy. For instance, for BERT training, the raw dataset is split into multiple small chunks. and for each of these chunks, we 
- read file in memory, perform tokenization, vocabulary lookup
- group sentences into documents according to document boundaries
- for each sentence in a document, generated random next sentence pairs and masks for language modeling
- prepare labels for next sentence prediction and masked language model
Then we have a dataset for samples suitable for BERT training.

Note that all the steps above happens before a dataset is formed (i.e. `transform()` cannot be used here). Performing all these steps with a single python process is slow. 

Now, the `DatasetStream.__iter__` API returns one dataset at a time, which is a problem if it is used in combination with `PrefetcherIter(worker_type='process')`, as the process-based prefetcher does not allow creation of child processes. 

## walk-arounds

1. Use `PrefetcherIter(worker_type='thread')`. It causes the training loop to slowdown because of the time sharing between the dataset creation thread and the thread for training the model. 
2. Use `PrefetcherIter(worker_type='process_no_daemon')`. It launches a process without daemon mode, where children processes can be created. This solves the speed issue, but at the cost of letting users maintain all these processes. In another word, users have call process.join() before exiting the program, otherwise the program hangs. 

## better solutions?

I think GluonNLP needs to provide a data loader for a `SimpleDatasetStream` that supports multiprocessing for dataset creation. I picked `SimpleDatasetStream` because it has a complete view of what data url to load, where `DatasetStream` does not provide that visibility. 

For implementation, the dataloading would happen in hierarchy using two worker pools. The first worker pool is responsible for loading and forming the dataset. The formed dataset is then passed to the second worker pool, where workers generate data batches that user wants. 

@leezu @szha

**Comment by eric-haibin-lin** [comment]
Also adding data loader expert @zhreshold

**Comment by zhreshold** [comment]
@eric-haibin-lin The best workaround IMO is to ditch the `PrefetchIter`, instead, let `SimpleDatasetStream` be self complete with option to prefetch dataset.

In this case, `SimpleDatasetStream` will have a daemon IO process (mapped to dataset list queue?, data queue, label queue, transform function queue) which do all the pre-processing tasks(reading, splitting...).
It will also have a process pool with each worker mapped to queues(data queue, label queue, transform function queue, output queue). Once IO process pushes splitted data/label/transform function to theses queues, they get process by all the workers in process pool. Main thread will grab the results from output queue.

Does it make sense? I know I am not fully aware of the whole pipeline, please correctly me if I am wrong.

**Comment by leezu** [comment]
So if I understand correctly the issue is that @eric-haibin-lin would like to spawn processes within the environment executing `DatasetStream.__iter__`.
Now this normally would be the standard python main process and there wouldn't be any issue; however, `PrefetcherIter` allows wrapping `DatasetStream.__iter__` into a separate (daemonic) process. Daemonic processes cannot spawn further processes in Python.

@zhreshold I'm not sure if extending SimpleDatasetStream with an option to prefetch dataset would solve the issue: It doesn't matter if SimpleDatasetStream or PrefetchIter is responsible for creating the separate (daemonic) process within which `DatasetStream.__iter__` is executed. In either case it is executed inside a daemonic process and cannot spawn further processes. Please let me know if I misunderstand your suggestion.

To avoid switching from daemonic to non-daemonic processes (for which the user must manually take care that they are correctly terminated on program termination), we may expose the already existing process pool to the user? (I'm not yet sure if this would lead to other issues) Alternatively we may provide an API where users can register their intent to use a process-pool and gluon-nlp may create the pool within the main process and pass the handle to the daemonic subprocess running `DatasetStream.__iter__`. Finally there may be an easy way to work around any issues caused by switching to non-daemonic processes.

**Comment by zhreshold** [comment]
My point is that daemon process can't have child process to accelerate tasks, however, it can have sibling processes which they will communicate by queues. 

The bad news is, it's not safe and elegent to expose queues in different classes, that's why I suggest to `DatasetStream` should be capable of prefetching in itself and all resources are encapsulated. 
Users will only touch `DatasetStream.__iter__` then.

**Comment by leezu** [comment]
Thanks for the clarification @zhreshold. In that case your proposed `MultiProcessingReadyDatasetStream` would expose an interface that would allow users to submit tasks to sibling processes during the execution of `DatasetStream.__iter__`?

I'd also like to add an orthogonal perspective: Besides using multiprocessing inside the `DatasetStream.__iter__` we may parallelise multiple calls to `DatasetStream.__iter__`. Currently we can only prefetch a single element from the `DatasetStream` in a single worker process (or thread). If multiple elements can be prefetched in parallel (possibly in different processes) we may achieve the same improved user-experience (lower latency).
This is currently not possible as `DatasetStream` is a Stream / python iterator and we can only access the next upcoming element. To retrieve multiple elements, eg. for distributing them to multiple processes for further processing, we have to call `next()` multiple times. But as `DatasetStream.__iter__` returns Dataset objects, each call to `next()` is expensive and in @eric-haibin-lin's case too slow.
Instead of directly constructing a Dataset on each call to `next()`, it may be sensible to work with [Futures](https://en.wikipedia.org/wiki/Futures_and_promises). Since Python 2 is in the process of being deprecated (though no concrete plan is made for GluonNLP yet), we may consider building on https://docs.python.org/3/library/concurrent.futures.html or https://docs.python.org/3/library/asyncio-task.html when addressing this problem. This may allow using existing standard library code and keeping our API simple. Thus, a call to `next()` would create a task and submit it to a pool of processes, but would not block calling `next()` again to begin computation for the following element.

**Comment by eric-haibin-lin** [comment]
@leezu thanks for the suggestion on `Future`s. Do we still have to change the DatasetStream interface if it is used?
